uint8 ww_code
string ww_str